### PR TITLE
Hardware Instancing fixes / sample (#1846)

### PR DESCRIPTION
**Pull request by mvaligursky**

### CODE REVIEW

**Comment by mvaligursky** [comment]
I have more complex Editor sample for hardware instancing as well.

**Comment by mvaligursky** [comment]
Note - I’ve just noticed the issue with culling - bounding box of original entity/mesh is used for culling. I need to disable culling on instanced objects.

File: examples/graphics/hardware-instancing.html [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            // copy matrix elements into array of floats
+            for (var m = 0; m < 16; m++)
+                matrices[matrixIndex++] = matrix.data[m];
+        }
+
+        // create static vertex buffer containing the matrices
+        var vertexBuffer = new pc.VertexBuffer(this.app.graphicsDevice, pc.VertexFormat.defaultInstancingFormat, instanceCount, pc.BUFFER_STATIC, matrices);
+
+        // initialise instancing using the vertex buffer on meshInstance of the created box
+        var boxMeshInst = box.model.model.meshInstances[0];
```
Comment (willeastcott) [inline examples/graphics/hardware-instancing.html]: Actually, you can access the `meshInstances` array on the `pc.ModelComponent`, so you can do the slightly simpler:

```var boxMeshInst = box.model.meshInstances[0];```

File: src/scene/mesh.js [hunk 2]
```diff
@@ -555,23 +555,51 @@ Object.assign(pc, function () {
...
-                this._buffer.setData(this.buffer);
+     /**
+      * @function
+      * @name pc.MeshInstance#setInstancing
+      * @description Sets up {@link pc.MeshInstance} to be rendered using Hardware Instancing.
+      * @param {pc.VertexBuffer|null} vertexBuffer - Vertex buffer to hold per-instance vertex data (usually world matrices).
+      * Pass null to turn off hardware instancing.
+      */
+    Object.assign(MeshInstance.prototype, {
+        setInstancing: function (vertexBuffer) {
```
Comment (willeastcott) [inline src/scene/mesh.js]: Can you define this in the previous `Object.assign(MeshInstance.prototype, {` block?

File: src/scene/mesh.js [hunk 3]
```diff
@@ -555,23 +555,51 @@ Object.assign(pc, function () {
...
             }
         }
     });
 
+    /**
+     * @name pc.MeshInstance#instancingCount
+     * @type {number}
+     * @description Number of instances when using hardware instancing to render the mesh.
+     */
+    Object.defineProperty(MeshInstance.prototype, 'instancingCount', {
```
Comment (willeastcott) [inline src/scene/mesh.js]: Same. Define in existing block to keep all `pc.MeshInstance` API toghether. 😄

**Comment by guycalledfrank** [comment]
Just saying - if I had time, I would implement instancing on top of Batch Groups, because it is related and also easy to grasp/use. Currently we have static (attach all meshes) and dynamic (attach and skin all meshes) batching. Now let's say we mark a group as "instanced" - engine would go over meshes, find cloned meshInstances, generate static per-instance vertex buffer, and that's it. We can also add some method to update this buffer.

**Comment by mvaligursky** [comment]
agreed, it's something I'm considering adding as well. This PR is basic functionality / examples allowing users to use it themselves, and we should integrate it with batching.

Even though I'd probably like to have batching dynamic .. current one has limitations - especially if you need to preserve sorting.

**Comment by guycalledfrank** [comment]
I tried to add completely dynamic instancing before, which would scan meshInstances to be rendered and do a single instanced draw instead - it didn't worth it, scanning through lots of objects and updating/creating buffers at runtime was too costly. This time I'd recommend generating preferably static intstancing groups and cull them with a combined AABB.

File: examples/graphics/hardware-instancing.html [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        app.setCanvasFillMode(pc.FILLMODE_FILL_WINDOW);
+        app.setCanvasResolution(pc.RESOLUTION_AUTO);
+
+        window.addEventListener("resize", function () {
+            app.resizeCanvas(canvas.width, canvas.height);
+        });
+
+        app.scene.ambientLight = new pc.Color(0.2, 0.2, 0.2);
+
+        // Create an Entity with a point light component and a sphere model component.
```
Comment (ellt92) [inline examples/graphics/hardware-instancing.html]: Doesn't look like there's a sphere model component used here.

File: src/graphics/device.js [hunk 5]
```diff
@@ -144,6 +144,12 @@ Object.assign(pc, function () {
      * @type {number}
      * @description The maximum supported texture anisotropy setting.
      */
+    /**
+     * @readonly
+     * @name pc.GraphicsDevice#extInstancing
+     * @type {boolean}
```
Comment (willeastcott) [inline src/graphics/device.js]: Actually, this isn't necessarily a boolean. For WebGL 2 it's a boolean. For WebGL 1, it's an object. This should probably be tidied up in the engine code at some point. BTW, we don't document the other pc.GraphicsDevice#extXYZ properties. Maybe we could introduce something like:

```
this.supportsInstancing = !!this.extInstancing;
```

There's already precedent for that with `supportsMsaa` and `supportsStencil`.

**Pull request merged**